File: deque/Deque.java
package deque;

@SuppressWarnings("unchecked")
public class Deque<T> {
    T[] deque;
    int front;
    int back;
    int size;
    int capacity;

    Deque() {
        front = 0;
        back = 0;
        size = 0;
        capacity = 3; // Default capacity of 3 just because.
        deque = (T[]) new Object[capacity];
    }

    private void rebuildDeque() {
        int newCapacity = capacity + capacity / 2; // 1.5x growth

        T[] newDeque = (T[]) new Object[newCapacity];

        int i = 0;
        int j = front;

        while (j != back) {
            newDeque[i] = deque[j];

            i++;
            j++;
            j %= capacity;
        }
        newDeque[i] = deque[back];

        capacity = newCapacity;
        deque = newDeque;
        front = 0;
        back = i;
    }

    public int getSize() {
        return size;
    }

    public int getCapacity() {
        return capacity;
    }

    public boolean isFull() {
        return size == capacity;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public T top() {
        return deque[front]; // returns null if empty
    }

    public T bottom() {
        return deque[back]; // returns null if empty
    }

    public void pushFront(T data) {
        if (isFull()) {
            rebuildDeque();
        } else if (isEmpty()) {
            front = 1;
        }

        front += capacity - 1; // n - 1
        front %= capacity; // = -1 (mod n)

        deque[front] = data;
        size++;
    }

    public void pushBack(T data) {
        if (isFull()) {
            rebuildDeque();
        } else if (isEmpty()) {
            back = -1;
        }

        back++;
        back %= capacity;

        deque[back] = data;
        size++;
    }

    public T popFront() {
        if (isEmpty()) {
            return null;
        }

        T data = deque[front];
        deque[front] = null;

        front++;
        front %= capacity;
        size--;

        return data;
    }

    public T popBack() {
        if (isEmpty()) {
            return null;
        }

        T data = deque[back];
        deque[back] = null;

        back += capacity - 1;
        back %= capacity;
        size--;

        return data;
    }

    public void clear() {
        deque = (T[]) new Object[capacity];
        size = 0;
        front = 0;
        back = 0;
    }

    public String toString() {
        if (isEmpty()) {
            return "";
        }

        StringBuilder result = new StringBuilder();

        int i = front;

        while (i != back) {
            if (deque[i] == null) {
                continue;
            } else {
                result.append(deque[i]).append(" ");
            }

            i++;
            i %= capacity;
        }

        result.append(deque[back]);

        return result.toString();
    }

    public static void main(String[] args) {
        Deque<Integer> intDeque = new Deque<>();

        for (int i = 0; i < 11; ++i) {
            int number = (7 * (i + 3)) % 11;
            System.out.printf("pushBack(%d)\n", number);
            intDeque.pushBack(number);
        }

        System.out.println("\nOriginal Deque");
        System.out.println("top(): " + intDeque.top());
        System.out.println("bottom(): " + intDeque.bottom());
        System.out.println("getSize(): " + intDeque.getSize());
        System.out.println("getCapacity(): " + intDeque.getCapacity());
        System.out.println("Deque: " + intDeque);

        intDeque.pushFront(17);
        System.out.println("\npushFront(17)");
        System.out.println("top(): " + intDeque.top());
        System.out.println("bottom(): " + intDeque.bottom());
        System.out.println("getSize(): " + intDeque.getSize());
        System.out.println("getCapacity(): " + intDeque.getCapacity());
        System.out.println("Deque: " + intDeque);

        intDeque.pushBack(12);
        System.out.println("\npushBack(12)");
        System.out.println("top(): " + intDeque.top());
        System.out.println("bottom(): " + intDeque.bottom());
        System.out.println("getSize(): " + intDeque.getSize());
        System.out.println("getCapacity(): " + intDeque.getCapacity());
        System.out.println("Deque: " + intDeque);

        intDeque.popFront();
        System.out.println("\npopFront()");
        System.out.println("top(): " + intDeque.top());
        System.out.println("bottom(): " + intDeque.bottom());
        System.out.println("getSize(): " + intDeque.getSize());
        System.out.println("getCapacity(): " + intDeque.getCapacity());
        System.out.println("Deque: " + intDeque);

        intDeque.popBack();
        System.out.println("\npopBack()");
        System.out.println("top(): " + intDeque.top());
        System.out.println("bottom(): " + intDeque.bottom());
        System.out.println("getSize(): " + intDeque.getSize());
        System.out.println("getCapacity(): " + intDeque.getCapacity());
        System.out.println("Deque: " + intDeque);

        intDeque.clear();
        System.out.println("\npopBack()");
        System.out.println("top(): " + intDeque.top());
        System.out.println("bottom(): " + intDeque.bottom());
        System.out.println("getSize(): " + intDeque.getSize());
        System.out.println("getCapacity(): " + intDeque.getCapacity());
        System.out.println("Deque: " + intDeque);
    }
}

File: deque/deque.c
#include "deque.h"

deque_t *create_deque(size_t type_size)
{
    deque_t *deque = malloc(sizeof(deque_t));

    deque->array = malloc(3 * type_size);
    deque->type_size = type_size;
    deque->size = 0;
    deque->capacity = 3;
    deque->front = 0;
    deque->back = 0;

    return deque;
}

void rebuild_deque(deque_t *deque)
{
    if( !deque )
    {
        return;
    }

    size_t new_capacity = deque->capacity + deque->capacity / 2;

    void *new_array = malloc(new_capacity * deque->type_size);

    memcpy(new_array,
           ( byte * )deque->array + deque->type_size * deque->front,
           deque->type_size * (deque->size - deque->front));

    if( deque->back < deque->front )
    {
        memcpy(( byte * )new_array +
                   deque->type_size * (deque->size - deque->front),
               deque->array,
               deque->type_size * deque->back);
    }

    free(deque->array);
    deque->array = new_array;
    deque->capacity = new_capacity;
    deque->front = 0;
    deque->back = deque->size - 1;
}

void delete_deque(deque_t **deque)
{
    if( !deque || !*deque )
    {
        return;
    }

    free((*deque)->array);
    free(*deque);
    *deque = nullptr;
}

void push_front(deque_t *deque, void *data)
{
    if( !deque || !data )
    {
        return;
    }

    if( deque->size == 0 && deque->front == deque->back )
    {
        deque->front = 1;
    }
    else if( deque->size == deque->capacity )
    {
        rebuild_deque(deque);
    }

    deque->front += deque->capacity - 1;
    deque->front %= deque->capacity;

    memcpy(( byte * )deque->array + deque->type_size * deque->front,
           data,
           deque->type_size);
    ++deque->size;
}

void push_back(deque_t *deque, void *data)
{
    if( !deque || !data )
    {
        return;
    }

    if( deque->size == 0 && deque->front == deque->back )
    {
        deque->back = -1;
    }
    else if( deque->size == deque->capacity )
    {
        rebuild_deque(deque);
    }

    ++deque->back;
    deque->back %= deque->capacity;

    memcpy(( byte * )deque->array + deque->type_size * deque->back,
           data,
           deque->type_size);
    ++deque->size;
}

void *top(deque_t *deque)
{
    if( !deque )
    {
        return nullptr;
    }

    return ( byte * )deque->array + deque->type_size * deque->front;
}

void *bottom(deque_t *deque)
{
    if( !deque )
    {
        return nullptr;
    }

    return ( byte * )deque->array + deque->type_size * deque->back;
}

bool pop_front(deque_t *deque, void *return_reference)
{
    if( !deque || deque->size == 0 )
    {
        return false;
    }

    void *data = ( byte * )deque->array + deque->type_size * deque->front;
    ++deque->front;
    deque->front %= deque->capacity;
    --deque->size;

    if( return_reference )
    {
        memcpy(return_reference, data, deque->type_size);
    }
    memset(data, 0, deque->type_size);

    return true;
}

bool pop_back(deque_t *deque, void *return_reference)
{
    if( !deque || deque->size == 0 )
    {
        return false;
    }

    void *data = ( byte * )deque->array + deque->type_size * deque->back;
    deque->back += deque->capacity - 1;
    deque->back %= deque->capacity;
    --deque->size;

    if( return_reference )
    {
        memcpy(return_reference, data, deque->type_size);
    }
    memset(data, 0, deque->type_size);

    return true;
}

void clear_deque(deque_t *deque)
{
    if( !deque )
    {
        return;
    }

    memset(deque->array, 0, deque->type_size * deque->capacity);
    deque->front = deque->back = deque->size = 0;
}
